symbolic-base/src/ZkFold/Symbolic/Data/Input.hs

```diff
@@ -0,0 +1,61 @@
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Symbolic.Data.Input (
+    SymbolicInput (..)
+) where
+
+import           Control.Monad.Representable.Reader (Rep)
+import           Data.Functor
+import           Data.Functor.Rep                   (Representable)
+import           Data.Ord                           (Ord)
+import           Data.Type.Equality                 (type (~))
+import           Data.Typeable                      (Proxy (..))
+import           GHC.Generics                       (Par1 (..))
+import           Prelude                            (fst, ($))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Data.ByteString        (Binary)
+import           ZkFold.Symbolic.Class
+import           ZkFold.Symbolic.Data.Bool
+import           ZkFold.Symbolic.Data.Class
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.MonadCircuit
+
+
+-- | A class for Symbolic input.
+class
+    ( SymbolicData d
+    , Support d ~ Proxy (Context d)
+    , Representable (Layout d)
+    , Binary (Rep (Layout d))
+    , Ord (Rep (Layout d))
+    ) => SymbolicInput d where
+    isValid :: d -> Bool (Context d)
+
+
+instance Symbolic c => SymbolicInput (Bool c) where
+  isValid (Bool b) = Bool $ fromCircuitF b $
+      \(Par1 v) -> do
+        u <- newAssigned (\x -> x v * (one - x v))
+        fst <$> runInvert (Par1 u)
```

Comment:
Maybe we can extract this `fst <$> runInvert` into a separate combinator to make it more readable? Like `isZero :: i -> m (Par1 i)`, for example